use_statement: Convert an exception to a future exception

eliransin opened the pull request:
The use statement execution code can throw if the keyspace is doesn't exist, this can be a problem for code that will use execute in a fiber since the exception will break the fiber even if `then_wrapped` is used.

Fixes #14449

mykaul commented:
Fixes?
Needs backport?

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2072/

avikivity commented:
@mykaul changing titles of single-patch requests has no effect, as the cover letter gets thrown away. See scripts/pull_github_pr.sh.

eliransin commented:
CI had spot termination, rerunning.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/2093/

avikivity commented:
> Fixes? Needs backport?

@eliransin ^

eliransin commented:
It doesn't fix anything in OSS but it fixes something in enterprise scylladb/scylla-enterprise#3082
Backports 5.3,5.2,5.1

mykaul commented:
> It doesn't fix anything in OSS but it fixes something in enterprise [scylladb/scylla-enterprise#3082](https://github.com/scylladb/scylla-enterprise/issues/3082)

It'd be great if you could open a clean description of the issue on OSS, and reference it in this PR.

eliransin commented:
> It'd be great if you could open a clean description of the issue on OSS, and reference it in this PR.

Done.
Should I do anything else? I'll reference the issue from the enterprise issue also.

mykaul commented:
@nyh , @avikivity  - please review.

nyh reviewed (commented):
I approve. But I am worried this may be masking a bigger bug that will bite us in the future. You are saying that if an statement::execute() implementation throws instead of returning an exception, the caller breaks. Why is this acceptable? Are we sure that no other statement's execute() implementation can throw in any case whatsoever?

Before I commit this patch I want to be sure we aren't missing something. Maybe it is the caller of execute() which needs to be fixed, not execute().

nyh commented:
@avikivity already approved this PR, but never merged it.
I said I approve too, but had a question, which the PR author (@eliransin) didn't answer.
In any case, a long time has passed and there are now conflicts, and this PR needs a rebase.
@eliransin, the ball is in your court.

eliransin commented:
Added the 'Fixes' tag to the commit message (not only cover letter) as it is a single patch PR.
Rebased the Patch, the only change was in the function prototype (which created the conflict)
Just in case compiled and ran again dev unit tests.

nyh commented:
> Added the 'Fixes' tag to the commit message (not only cover letter) as it is a single patch PR. Rebased the Patch, the only change was in the function prototype (which created the conflict) Just in case compiled and ran again dev unit tests.

Thanks. I'll merge this (after it passes CI!) even though I still don't understand what problem it solves. In other words, we can add try/catch in a thousand places in the code (I even have in tutorial.md a section on "Exceptions vs. exceptional futures" that discusses this topic, without a clear conclusion), but how did you decide to add it in **this** place? That's the only thing that's not clear to me about the patch - the code itself of the patch is obviously trivial.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4598/
- Duration: 6 hr 30 min
- Builder: monster

nyh commented:
> Thanks. I'll merge this (after it passes CI!) 

It passed CI so I'll merge it now. @kostja will be happy - ABC ;-)

> even though I still don't understand what problem it solves.

I read again the discussion above, and realized it solves a problem that only exists in the Scylla Enterprise branch, in its audit code. I think it means that audit code is buggy and makes unnecessary assumptions on statements it executes - that they can't throw. I also wonder if "USE" is really the only statement that can ever throw instead of returning a future. But if this patch helps then, ok, I guess.

bhalevy commented on cql3/statements/use_statement.cc:
```diff
@@ -56,7 +56,11 @@ future<> use_statement::check_access(query_processor& qp, const service::client_
+    try {
+        state.get_client_state().set_keyspace(qp.db().real_database(), _keyspace);
+    } catch(...) {
+        return make_exception_future<::shared_ptr<cql_transport::messages::result_message>>(std::current_exception());
```
Hi, I have 2 FYI's regarding this change that I just saw:
1. You could have used `current_exception_as_future<::shared_ptr<cql_transport::messages::result_message>>()`
2. You could turn this function into a coroutine.

And a potential problem:
3. can't `::make_shared` below throw too?  (2. above would fix hande that and return an exception future in all cases)

It might be fine to throw on "environmental" failures and return a failed future on "business logic" errors, but it's harder to handle errors this way.
